### PR TITLE
Add support for EventBridge s3 events in lambda-promtail

### PR DESCRIPTION
**Pull request by thepalbi**
**What this PR does / why we need it**:
Lambda promtail supports s3 events, which are used for scraping several log sources such as ALB access logs. This works by configuring at the S3 bucket level "s3 event notification", that are configured to target the lambda deployment of lambda-promtail.

However, if one is configuring this through CloudFormation, there's a known issue with AWS that doesn't allow to configure both the lambda, the bucket, and the notifications in the same stack. See [this issue](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/79) for details.

For that, AWS introduced EventBridge notifications, which can be used to ship s3 events to a lambda deployment as well. This flow looks like:
s3 -> eventbridge bus -> eventbridge rule -> lambda

EventBridge has it's own message structure for s3 notifications. This PR adds a translation layer, just for `Object created` events (since they are the only ones we should take into account), so that EventBridge events can be received, and trigger the lambda as if they were from s3.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/10209

**Special notes for your reviewer**:

- [x] Pending testing this with an actual deployment of the s3 -> event bridge -> lambda flow
- [x] ~~Add CF template for the `s3 -> event bridge -> lambda` deployment~~ Follow up PR

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: tools/lambda-promtail/lambda-promtail/eventbridge.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	if !(ev.Source == "aws.s3" && ev.DetailType == "Object Created") {
+		return fmt.Errorf("event bridge event type not supported")
+	}
+
+	var eventDetail S3Detail
+	if err := json.Unmarshal(ev.Detail, &eventDetail); err != nil {
+		return err
+	}
+
+	// TODO(thepalbi): how to fill bucket owner?
```
Comment (thepalbi) [inline tools/lambda-promtail/lambda-promtail/eventbridge.go:43]: the current logic extracts [this other field](https://github.com/grafana/loki/blob/4cd76919100fe59d41bd832bb07518f267d4ea1d/tools/lambda-promtail/lambda-promtail/s3.go#L208) from the original s3 events (the owner's principal ID), but this ends up not being used. See [here](https://github.com/grafana/loki/blob/4cd76919100fe59d41bd832bb07518f267d4ea1d/tools/lambda-promtail/lambda-promtail/s3.go#L248).

EventBridge events don't include the principal ID, but include the bucket AWS acct id, which is the value that should be used in [here](https://github.com/grafana/loki/blob/4cd76919100fe59d41bd832bb07518f267d4ea1d/tools/lambda-promtail/lambda-promtail/s3.go#L248):

See doc below in the field where it was supposed to be used in:
```
	// The account ID of the expected bucket owner. If the bucket is owned by a
	// different account, the request will fail with an HTTP 403 (Access Denied) error.
	ExpectedBucketOwner *string
```

**Comment by cstyan** [comment]
@thepalbi should we provide updated terraform and cloudformation examples with this?

**Comment by thepalbi** [comment]
> @thepalbi should we provide updated terraform and cloudformation examples with this?

So the CloudFormation/terraform templates we provide are just for configuring the basic use case of deploying the lambda, IAM roles and a CloudWatch subscription filter (which makes AWS forward logs from a CW Log group into the Lambda).

This PR is just for tackling a really specific use case, which is allowing the lambda to process logs from s3 (which for example, is the only method of logging for application load balancer access logs), using EventBridge as the middle man to notify the lambda when a file get's added to S3. EventBridge is actually and extra-hop, to tackle a limitation CloudFormation has when defining where a bucket should send notifications. I'm almost certain no such limitation exists in Terraform, but let me double check.

Either way, I think it would be nice to have a separate CloudFormation template for this use case, since it's a known problem. I'll do the testing I have to do with CloudFormation, and add the template in the PR.

**Review by cstyan — APPROVED** [review]
We can also add examples for tf and cf in a follow up PR, I'll let you decide.

**Comment by thepalbi** [comment]
@cstyan tested on a real scenario with: `application LB -> access logging -> s3 -> notification to EventBridge -> EB -> lambda promtail -> Loki`, and working as expected:

image

[Here's](https://github.com/grafana/loki/compare/pablo/lambda-promtail-event-bridge-support...pablo/lambda-promtail-event-bridge-setup?expand=1) the edited CF template, which I'll add as a separate one, alongside with some docs, in a follow up PR.

**Pull request merged**